src/Storages/Statistics/StatisticsCountMinSketch.cpp

```diff
@@ -0,0 +1,111 @@
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <Storages/Statistics/StatisticsCountMinSketch.h>
+#include <Interpreters/convertFieldToType.h>
+#include <DataTypes/FieldToDataType.h>
+
+#if USE_DATASKETCHES
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_STATISTICS;
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+/// Constants chosen based on rolling dices, which provides an error tolerance of 0.1% (ε = 0.001) and a confidence level of 99.9% (δ = 0.001).
+/// And sketch the size is 152kb.
+static constexpr auto num_hashes = 7uz;
+static constexpr auto num_buckets = 2718uz;
+
+StatisticsCountMinSketch::StatisticsCountMinSketch(const SingleStatisticsDescription & stat_, DataTypePtr data_type_)
+    : IStatistics(stat_)
+    , sketch(num_hashes, num_buckets)
+    , data_type(data_type_)
+{
+}
+
+Float64 StatisticsCountMinSketch::estimateEqual(const Field & val) const
+{
+    if (data_type->isValueRepresentedByNumber())
+    {
+        /// 'val' maybe number or string, method 'convertFieldToType' will
+        ///     1. convert string to number, date, datetime, IPv4, Decimal etc
+        ///     2. return null if val larger than the range of data_type
+        auto val_converted = convertFieldToType(val, *data_type);
+        if (val_converted.isNull())
+            return 0;
+
+        /// We will get the proper data type of val_converted, for example, UInt8 for 1, UInt16 for 257.
+        auto data_type_converted = applyVisitor(FieldToDataType<LeastSupertypeOnError::Null>(), val_converted);
+        DataTypes data_types = {data_type, data_type_converted};
+        auto super_type = tryGetLeastSupertype(data_types);
+
+        /// If data_type is UInt8 but val_typed is UInt16, we should return 0.
+        if (!super_type->equals(*data_type))
+            return 0;
+
+        return sketch.get_estimate(&val_converted, data_type->getSizeOfValueInMemory());
+    }
+
+    if (isStringOrFixedString(data_type))
+    {
+        return sketch.get_estimate(val.get<String>());
+    }
+
+    throw Exception(
```

Comment:
is it possible to write a test covering this exception?

Reply:
The code will never be reached in logical, it is just for sementic usage.

Reply:
Then please throw LOGICAL_ERROR